vllm/distributed/kv_transfer/kv_connector/v1/nixl_connector.py

```diff
@@ -0,0 +1,647 @@
+        # TODO(tms): self.block_len needs to be per-layer for sliding window,
+        # hybrid attn, etc
+        self.block_len = kv_elem_size * math.prod(first_kv_cache.shape[-3:])
+
+        logger.debug("Per layer kv cache size: %s", first_kv_cache[0].shape)
+        self.num_blocks = num_blocks
+        self.dst_num_blocks[self.engine_id] = num_blocks
+        self.kv_caches = kv_caches
+        kv_caches_base_addr = []
+        caches_data = []
+
+        # Note(tms): I modified this from the original region setup code.
+        # K and V are now in different regions. Advantage is that we can
+        # elegantly support MLA and any cases where the K and V tensors
+        # are non-contiguous (it's not locally guaranteed that they will be)
+        # Disadvantage is that the encoded NixlAgentMetadata is now larger
+        # (roughly 8KB vs 5KB).
+        for layer_name in kv_caches:
+            for cache in kv_caches[layer_name]:
+                base_addr = cache.data_ptr()
+                region_len = num_blocks * self.block_len
+                caches_data.append((base_addr, region_len, self.rank, ""))
+                kv_caches_base_addr.append(base_addr)
+        self.kv_caches_base_addr[self.engine_id] = kv_caches_base_addr
+        self.num_regions = len(caches_data)
+
+        descs = self.nixl_wrapper.get_reg_descs(caches_data, "VRAM")
+        logger.debug("Registering descs: %s", caches_data)
+        self.nixl_wrapper.register_memory(descs)
+        logger.debug("Done registering descs")
+
+        self._registered_descs.append(descs)
+
+        # THIS IS FOR DEV
+        _ctx = zmq.Context()  # type: ignore
+        _side_channel = _ctx.socket(zmq.PAIR)  # type: ignore
+        NIXL_ROLE = os.getenv("NIXL_ROLE")
+
+        # FOR DEV, SENDER puts data in its KV caches so the RECVER can check it
+        if os.environ.get("VLLM_DEBUG_INITIAL_NIXL_PD_XFER") is not None:
+            n_blocks_to_send = min(4096, kv_caches[first_layer_name].shape[1])
+            debug_xfer_gb = 2.0 * n_blocks_to_send * self.block_len / 1e9
+            logger.debug(
+                "Starting initial NIXL PD XFER: Total %s GB, Block len %s KB",
+                debug_xfer_gb, self.block_len / 1024)
+            if NIXL_ROLE == "SENDER":
+                for b in range(n_blocks_to_send):
+                    kv_caches[first_layer_name][0, b, 0, 0, 0] = b + 100.0
+                    kv_caches[first_layer_name][1, b, 0, 0, 0] = b + 200.0
+
+            for b in range(5):
+                logger.debug("%s KV_CACHE coord %s val %f", NIXL_ROLE,
+                             (0, b, 0, 0, 0),
+                             kv_caches[first_layer_name][0, b, 0, 0, 0])
+                logger.debug("%s KV_CACHE coord %s val %f", NIXL_ROLE,
+                             (1, b, 0, 0, 0),
+                             kv_caches[first_layer_name][1, b, 0, 0, 0])
+            remote_engine_id = None  # HACK for debug send
+
+        if NIXL_ROLE == "SENDER":
```

Comment:
Should this be also be nested under the dev if above?

Reply:
No, this is still needed for handshaking